File: scripts/stream_logs.mjs
/**
 * Aegis Protocol — JSON-RPC Streaming Log Viewer
 * 
 * Watches logs/stream.log and renders each JSON-RPC entry with
 * color-coded severity levels in real-time.
 * 
 * Usage:
 *   node scripts/stream_logs.mjs
 *   node scripts/stream_logs.mjs --follow     (tail -f mode, default)
 *   node scripts/stream_logs.mjs --replay     (replay existing log then follow)
 *   node scripts/stream_logs.mjs --replay-only (replay existing log and exit)
 * 
 * Severity Colors:
 *   info     → Cyan
 *   warn     → Yellow
 *   error    → Red
 *   critical → White on Red background
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const ROOT = path.resolve(__dirname, '..');
const LOG_PATH = path.join(ROOT, 'logs', 'stream.log');

// ─── ANSI Colors ────────────────────────────────────────────────────
const C = {
  reset:    '\x1b[0m',
  bold:     '\x1b[1m',
  dim:      '\x1b[2m',
  red:      '\x1b[31m',
  green:    '\x1b[32m',
  yellow:   '\x1b[33m',
  blue:     '\x1b[34m',
  magenta:  '\x1b[35m',
  cyan:     '\x1b[36m',
  white:    '\x1b[37m',
  bgRed:    '\x1b[41m',
  bgGreen:  '\x1b[42m',
  bgYellow: '\x1b[43m',
  bgBlue:   '\x1b[44m',
  bgMagenta:'\x1b[45m',
};

// ─── Severity Styling ───────────────────────────────────────────────
const SEVERITY_STYLES = {
  info:     { icon: 'ℹ', color: C.cyan,   bg: '' },
  warn:     { icon: '⚠', color: C.yellow, bg: '' },
  error:    { icon: '✖', color: C.red,    bg: '' },
  critical: { icon: '🚨', color: C.white,  bg: C.bgRed },
};

// ─── Method Category Colors ─────────────────────────────────────────
function getMethodColor(method) {
  if (method.includes('telecom'))    return C.blue;
  if (method.includes('deepfake'))   return C.magenta;
  if (method.includes('mule'))       return C.yellow;
  if (method.includes('adjudicate')) return C.cyan;
  if (method.includes('guard'))      return C.red;
  if (method.includes('mha'))        return C.green;
  if (method.includes('pipeline'))   return C.green;
  if (method.includes('flag'))       return C.yellow;
  if (method.includes('cleared'))    return C.green;
  return C.white;
}

// ─── Render a single log entry ──────────────────────────────────────
function renderLogEntry(line) {
  try {
    const entry = JSON.parse(line.trim());
    
    if (!entry.jsonrpc || entry.jsonrpc !== '2.0') {
      // Not a JSON-RPC entry, print raw
      console.log(`${C.dim}${line.trim()}${C.reset}`);
      return;
    }

    const level = entry.params?._level || 'info';
    const style = SEVERITY_STYLES[level] || SEVERITY_STYLES.info;
    const methodColor = getMethodColor(entry.method || '');
    const ts = entry.params?._timestamp || new Date().toISOString();

    // Build the formatted line
    const severityTag = style.bg
      ? `${style.bg}${style.color}${C.bold} ${style.icon} ${level.toUpperCase()} ${C.reset}`
      : `${style.color}${style.icon} ${level.toUpperCase().padEnd(8)}${C.reset}`;

    const methodTag = `${methodColor}${C.bold}${entry.method || 'unknown'}${C.reset}`;
    const idTag = `${C.dim}${entry.id || ''}${C.reset}`;
    const timeTag = `${C.dim}${ts.substring(11, 23)}${C.reset}`;

    console.log(`${timeTag} ${severityTag} ${methodTag} ${idTag}`);

    // Print key params (excluding internal fields)
    const params = { ...entry.params };
    delete params._timestamp;
    delete params._level;

    if (Object.keys(params).length > 0) {
      const paramsStr = JSON.stringify(params, null, 2)
        .split('\n')
        .map(l => `  ${C.dim}│${C.reset} ${l}`)
        .join('\n');
      console.log(paramsStr);
    }

    console.log(`  ${C.dim}└${'─'.repeat(60)}${C.reset}`);

  } catch {
    // Not valid JSON, print raw
    if (line.trim()) {
      console.log(`${C.dim}${line.trim()}${C.reset}`);
    }
  }
}

// ─── Main ───────────────────────────────────────────────────────────
function main() {
  const args = process.argv.slice(2);
  const replayOnly = args.includes('--replay-only');
  const replay = args.includes('--replay') || replayOnly;
  const follow = !replayOnly;

  console.log('');
  console.log(`${C.bold}${C.cyan}╔══════════════════════════════════════════════════════════════╗${C.reset}`);
  console.log(`${C.bold}${C.cyan}║   🛡️  AEGIS PROTOCOL — JSON-RPC STREAM VIEWER              ║${C.reset}`);
  console.log(`${C.bold}${C.cyan}║   Log: logs/stream.log                                      ║${C.reset}`);
  console.log(`${C.bold}${C.cyan}║   Mode: ${(replay ? (replayOnly ? 'Replay Only' : 'Replay + Follow') : 'Follow (live)').padEnd(52)}║${C.reset}`);
  console.log(`${C.bold}${C.cyan}╚══════════════════════════════════════════════════════════════╝${C.reset}`);
  console.log('');

  // Ensure log directory exists
  const logsDir = path.dirname(LOG_PATH);
  if (!fs.existsSync(logsDir)) {
    fs.mkdirSync(logsDir, { recursive: true });
  }

  // Ensure log file exists
  if (!fs.existsSync(LOG_PATH)) {
    fs.writeFileSync(LOG_PATH, '');
  }

  // Replay existing entries
  if (replay) {
    const existing = fs.readFileSync(LOG_PATH, 'utf-8');
    const lines = existing.split('\n').filter(l => l.trim());
    
    if (lines.length > 0) {
      console.log(`${C.dim}─── Replaying ${lines.length} existing entries ───${C.reset}`);
      for (const line of lines) {
        renderLogEntry(line);
      }
      console.log(`${C.dim}─── End of replay ───${C.reset}`);
      console.log('');
    } else {
      console.log(`${C.dim}─── No existing entries to replay ───${C.reset}`);
      console.log('');
    }
  }

  if (!follow) {
    console.log(`${C.green}Replay complete.${C.reset}`);
    return;
  }

  // Watch for new entries
  console.log(`${C.cyan}Watching for new log entries... (Ctrl+C to stop)${C.reset}`);
  console.log('');

  let fileSize = fs.statSync(LOG_PATH).size;

  fs.watchFile(LOG_PATH, { interval: 100 }, () => {
    try {
      const stats = fs.statSync(LOG_PATH);
      if (stats.size > fileSize) {
        // Read only the new data
        const fd = fs.openSync(LOG_PATH, 'r');
        const buffer = Buffer.alloc(stats.size - fileSize);
        fs.readSync(fd, buffer, 0, buffer.length, fileSize);
        fs.closeSync(fd);

        const newData = buffer.toString('utf-8');
        const newLines = newData.split('\n').filter(l => l.trim());

        for (const line of newLines) {
          renderLogEntry(line);
        }

        fileSize = stats.size;
      } else if (stats.size < fileSize) {
        // Log file was truncated/reset
        fileSize = 0;
        console.log(`${C.yellow}⚠ Log file reset detected${C.reset}`);
      }
    } catch {
      // File might be temporarily unavailable
    }
  });

  // Keep process alive
  process.on('SIGINT', () => {
    console.log(`\n${C.dim}Stream viewer stopped.${C.reset}`);
    fs.unwatchFile(LOG_PATH);
    process.exit(0);
  });
}

main();

File: scripts/stream_logs.py
#!/usr/bin/env python3
import time
import json
import os
import sys

if sys.stdout.encoding.lower() != 'utf-8':
    try:
        sys.stdout.reconfigure(encoding='utf-8')
    except AttributeError:
        pass


# Color constants
class C:
    RESET = '\033[0m'
    BOLD = '\033[1m'
    DIM = '\033[2m'
    RED = '\033[31m'
    GREEN = '\033[32m'
    YELLOW = '\033[33m'
    BLUE = '\033[34m'
    MAGENTA = '\033[35m'
    CYAN = '\033[36m'
    WHITE = '\033[37m'
    BG_RED = '\033[41m'

SEVERITY_STYLES = {
    'info': {'icon': 'ℹ', 'color': C.CYAN, 'bg': ''},
    'warn': {'icon': '⚠', 'color': C.YELLOW, 'bg': ''},
    'error': {'icon': '✖', 'color': C.RED, 'bg': ''},
    'critical': {'icon': '🚨', 'color': C.WHITE, 'bg': C.BG_RED},
}

def get_method_color(method):
    if 'telecom' in method: return C.BLUE
    if 'deepfake' in method: return C.MAGENTA
    if 'mule' in method: return C.YELLOW
    if 'adjudicate' in method: return C.CYAN
    if 'guard' in method: return C.RED
    if 'mha' in method: return C.GREEN
    if 'pipeline' in method: return C.GREEN
    if 'flag' in method: return C.YELLOW
    if 'cleared' in method: return C.GREEN
    return C.WHITE

def render_log_entry(line):
    line = line.strip()
    if not line: return
    try:
        entry = json.loads(line)
        if entry.get('jsonrpc') != '2.0':
            print(f"{C.DIM}{line}{C.RESET}")
            return
            
        params = entry.get('params', {})
        level = params.get('_level', 'info')
        style = SEVERITY_STYLES.get(level, SEVERITY_STYLES['info'])
        method = entry.get('method', 'unknown')
        method_color = get_method_color(method)
        ts = params.get('_timestamp', '')
        
        severity_tag = f"{style['bg']}{style['color']}{C.BOLD} {style['icon']} {level.upper()} {C.RESET}" if style['bg'] else f"{style['color']}{style['icon']} {level.upper().ljust(8)}{C.RESET}"
        method_tag = f"{method_color}{C.BOLD}{method}{C.RESET}"
        id_tag = f"{C.DIM}{entry.get('id', '')}{C.RESET}"
        time_tag = f"{C.DIM}{ts[11:23]}{C.RESET}" if ts else ""
        
        print(f"{time_tag} {severity_tag} {method_tag} {id_tag}")
        
        # Print params
        clean_params = {k: v for k, v in params.items() if not k.startswith('_')}
        if clean_params:
            param_str = json.dumps(clean_params, indent=2)
            for p_line in param_str.split('\n'):
                print(f"  {C.DIM}│{C.RESET} {p_line}")
        print(f"  {C.DIM}└{'─'*60}{C.RESET}")
        
    except json.JSONDecodeError:
        print(f"{C.DIM}{line}{C.RESET}")

def main():
    log_path = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'logs', 'stream.log')
    
    print(f"\n{C.BOLD}{C.CYAN}╔══════════════════════════════════════════════════════════════╗{C.RESET}")
    print(f"{C.BOLD}{C.CYAN}║   🛡️  AEGIS PROTOCOL — JSON-RPC STREAM VIEWER              ║{C.RESET}")
    print(f"{C.BOLD}{C.CYAN}║   Log: logs/stream.log                                      ║{C.RESET}")
    print(f"{C.BOLD}{C.CYAN}╚══════════════════════════════════════════════════════════════╝{C.RESET}\n")

    os.makedirs(os.path.dirname(log_path), exist_ok=True)
    if not os.path.exists(log_path):
        open(log_path, 'w').close()

    if '--follow' not in sys.argv:
        # Replay only
        with open(log_path, 'r') as f:
            for line in f:
                render_log_entry(line)
        return

    # Follow mode
    print(f"{C.CYAN}Watching for new log entries... (Ctrl+C to stop){C.RESET}\n")
    with open(log_path, 'r') as f:
        f.seek(0, 2) # Go to end
        while True:
            line = f.readline()
            if not line:
                time.sleep(0.1)
                continue
            render_log_entry(line)

if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print(f"\n{C.DIM}Stream viewer stopped.{C.RESET}")
        sys.exit(0)
